File: algorithm/balanced_binary_tree/balanced_binary_tree.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	ld := depth(root.Left)
	rd := depth(root.Right)
	if ld == -1 || rd == -1 || abs(ld-rd) > 1 {
		return false
	}
	return true
}

func depth(node *TreeNode) int {
	if node == nil {
		return 0
	}
	ld := depth(node.Left)
	rd := depth(node.Right)
	if ld == -1 || rd == -1 || abs(ld-rd) > 1 {
		return -1
	}
	return max(ld, rd) + 1
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	n11 := TreeNode{11, nil, nil}
	n10 := TreeNode{10, nil, nil}
	n9 := TreeNode{9, nil, nil}
	n8 := TreeNode{8, nil, nil}
	n7 := TreeNode{7, nil, nil}
	n6 := TreeNode{6, nil, nil}
	n5 := TreeNode{5, nil, nil}
	n4 := TreeNode{4, nil, nil}
	n3 := TreeNode{3, nil, nil}
	n2 := TreeNode{2, nil, nil}
	n1 := TreeNode{1, nil, nil}
	n3.Right = &n7
	n3.Left = &n6
	n2.Right = &n5
	n2.Left = &n4
	n1.Right = &n3
	n1.Left = &n2
	fmt.Printf("Is Tree Balanced? Expect true, Get: %v\n", isBalanced(&n1))
	n7.Right = &n8
	n8.Right = &n9
	n9.Right = &n10
	fmt.Printf("Is Tree Balanced? Expect false, Get: %v\n", isBalanced(&n1))
	fmt.Printf("Is Tree Balanced? Expect true, Get: %v\n", isBalanced(&n11))
	fmt.Printf("Is Tree Balanced? Expect true, Get: %v\n", isBalanced(nil))
}

File: algorithm/balanced_binary_tree/cpp/balanced_binary_tree.cpp
#include <iostream>
#include <vector>

using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        int ld = depth(root->left);
        int rd = depth(root->right);
        if (ld == -1 || rd == -1 || abs(rd-ld) > 1) {
            return false;
        }
        return true;
        
    }
    
    int depth(TreeNode* node) {
        if (node == NULL) {
            return 0;
        }
        int ld = depth(node->left);
        int rd = depth(node->right);
        if (ld == -1 || rd == -1 || abs(rd-ld) > 1) {
            return -1;
        }
        return ld > rd ? ld+1 : rd+1;
    }
    
    int abs(int x) {
        if (x < 0) {
            return -x;
        }
        return x;
    }
};
